internal/lsp/hover.go

```diff
@@ -0,0 +1,174 @@
+package lsp
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/olekukonko/tablewriter"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/types"
+)
+
+type BuiltinPosition struct {
+	Builtin *ast.Builtin
+	Line    uint
+	Start   uint
+	End     uint
+}
+
+var builtins = builtinMap() //nolint:gochecknoglobals
+
+func builtinMap() map[string]*ast.Builtin {
+	m := make(map[string]*ast.Builtin)
+	for _, b := range ast.CapabilitiesForThisVersion().Builtins {
+		m[b.Name] = b
+	}
+
+	return m
+}
+
+func builtinCategory(builtin *ast.Builtin) (category string) {
+	if len(builtin.Categories) == 0 {
+		if s := strings.Split(builtin.Name, "."); len(s) > 1 {
+			category = s[0]
+		} else {
+			category = builtin.Name
+		}
+	} else {
+		category = builtin.Categories[0]
+	}
+
+	return category
+}
+
+func writeFunctionSnippet(sb *strings.Builder, builtin *ast.Builtin) {
+	sb.WriteString("```rego\n")
+
+	resType := builtin.Decl.NamedResult()
+	if n, ok := resType.(*types.NamedType); ok {
+		sb.WriteString(n.Name)
+	} else {
+		sb.WriteString("output")
+	}
+
+	sb.WriteString(" := ")
+
+	sb.WriteString(builtin.Name)
+	sb.WriteString("(")
+
+	for i, arg := range builtin.Decl.NamedFuncArgs().Args {
+		if i > 0 {
+			sb.WriteString(", ")
+		}
+
+		if n, ok := arg.(*types.NamedType); ok {
+			sb.WriteString(n.Name)
+		} else {
+			sb.WriteString(arg.String())
+		}
+	}
+
+	sb.WriteString(")\n```")
+}
+
+func createHoverContent(builtin *ast.Builtin) string {
```

Comment:
It might be nice to have a simple test case for this function to make for easier refactoring. E.g. we might want to make use of a template file or similar in future.

Reply:
Sounds good. I'll look into adding a couple of tests.